### PR TITLE
Delete focus upstream

### PR DESCRIPTION
**Pull request by geetanshu-m**
### Summary of Changes

Following features has been added:

- Added Delete Focus button in the Smart-Crop plugin.
- Added the methods for deleting the focus points and resized images for a particular image.
- Now, If the original image is deleted it's focus point and it's resized images would be deleted on next time the article (containing the image) would be loaded.
- Location of resized images has been changed from ```/images/.cache``` to ```/media/focus```.
- Adding ```<picture>``` element to only those images whose focus are being set.
- Image quality field has been removed from the Smart-Crop plugin.

### Testing Instructions

- Add image to the article, set it's focus points (useing focus editor).
- Now try to delete the image, load the same article again, ```<picture>``` tag whould be removed from the image and the resized images of that image whould be removed from ```/media/focus```.

##### For deleteing the focus and resized images from focus editor.

- In case you want to delete the focus and the previously resized images.
- Go to the focus editor and hit ```Delete Focus``` button, and hit cancel.
- All the resized images of that image whould be removed from ```/media/focus```.

##### New Look of the plugin
screenshot 28

### CODE REVIEW

File: media/plg_media-action_smartcrop/js/smartcrop.js [hunk 1]
```diff
@@ -114,12 +114,35 @@ Joomla.MediaManager.Edit = Joomla.MediaManager.Edit || {};
...
+
+		var deleteButton = document.createElement("BUTTON");
+		deleteButton.setAttribute("type", "button");
+		deleteButton.setAttribute("id", "delete-focus");
+		deleteButton.setAttribute("class", "btn btn-danger");
+		deleteButton.appendChild(document.createTextNode("Delete Focus"));
+		deleteButton.addEventListener('click', function() {
+			Joomla.request({
+				url: resolveBaseUrl() + "/administrator/index.php?option=com_media&task=adaptiveimage.deleteFocus&path=" + path,
+				method: 'GET',
```
Comment (kasvith) [inline media/plg_media-action_smartcrop/js/smartcrop.js]: can't we go for a `POST` here?

Comment (geetanshu-m) [reply media/plg_media-action_smartcrop/js/smartcrop.js]: ```POST``` will work fine here, but I guess ```GET``` requests are faster.

Comment (kasvith) [reply media/plg_media-action_smartcrop/js/smartcrop.js]: Use `GET` for receive information, Use `POST` to commit an action, we can't use DELETE in here as some servers may not support it. Always when committing an action try something secure like POST.

On the server's side, they are only just a request

**Comment by kasvith** [comment]
Will do a manual test on this, seems fine for me

File: administrator/components/com_media/src/Controller/AdaptiveImageController.php [hunk 2]
```diff
@@ -65,6 +65,19 @@ public function execute($task)
 				$widths = json_decode($this->input->getString('widths'), true);
 				$this->cropImage($imgPath, $widths);
 				return true;
+			case "deleteFocus" :
+				if ($this->input->getMethod() != 'POST')
+				{
+					return false;
+				}
+				else
+				{
```
Comment (nibra) [inline administrator/components/com_media/src/Controller/AdaptiveImageController.php]: This "else" is not needed.

File: libraries/src/AdaptiveImage/JSONFocusStore.php [hunk 3]
```diff
@@ -131,6 +138,60 @@ public function getFocus($imgPath, $width = null)
...
+	 * 
+	 * @param   string  $imgSrc  Path of the image
+	 * 
+	 * @return  boolean
+	 * 
+	 * @since 4.0.0
+	 */
+	public function deleteFocus($imgSrc)
+	{
+		if (filesize(static::$dataLocation))
```
Comment (nibra) [inline libraries/src/AdaptiveImage/JSONFocusStore.php]: You mean
```php
    if (filesize(static::$dataLocation) > 0)
```
?

**Review by nibra — CHANGES_REQUESTED** [review]
Only few changes needed before merge.
Please set on the todo list to use a layout for creation of the pictutre element (future improvement).

File: libraries/src/AdaptiveImage/JSONFocusStore.php (hunk 3 above)
Comment (geetanshu-m) [reply libraries/src/AdaptiveImage/JSONFocusStore.php]: Yes

Comment (kasvith) [reply libraries/src/AdaptiveImage/JSONFocusStore.php]: use `> 0`

**Comment by kasvith** [comment]
Good work Geetanshu

**Pull request merged**